Patch/1.7/merge 1.6.23

AlbsAgren opened the pull request.

AlbsAgren committed:
service: always send metric event on server exit

Before: if a server exited during a call and wasnt able to send an ACK,
there would be no metric for the cancelled service call.
Now: we always send a metric event if a server exits.

Resolves #486

AlbsAgren committed:
changelog

AlbsAgren committed:
Merge tag '1.6.23' into patch/1.7/merge-1.6.23

1.6.23

lazan commented on middleware/service/source/manager/handle.cpp:
```diff
@@ -69,16 +69,28 @@ namespace casual
+                     metric.service = caller.service;
+                     metric.process = callee;
+
+                     state.metric.add( metric);
```
`std::move( metric)`

lazan commented on middleware/service/include/service/manager/state.h:
```diff
@@ -97,6 +97,7 @@ namespace casual
                common::process::Handle process;
                common::strong::correlation::id correlation;
                common::transaction::global::ID gtrid;
+               std::string service;
```
use `state::service::id::type` instead?

lazan replied:
And remove `m_reserved_service` and use `caller().service` to get _reserved-service_. i think this will be an improvement.

AlbsAgren replied:
Had a go at it, maybe a little awkward on TPNOREPLY where caller is empty? Have a look.

lazan commented on middleware/service/source/manager/handle.cpp:
```diff
@@ -69,16 +69,28 @@ namespace casual
+                     metric.code.result = code;
+                     metric.correlation = caller.correlation;
+                     metric.service = caller.service;
+                     metric.process = callee;
```
I want us to provide the transaction id, if we got it.
In `Caller` we got `gtrid`. We could just create an transaction::ID from `gtrid` with no `bqual`, or keep a `transaction::ID` in `Caller` instead of `transaction::global::ID`

lazan commented on middleware/service/source/manager/handle.cpp:
```diff
@@ -278,7 +290,7 @@ namespace casual
                                  event.state.pid, " - caller: ", caller.process.pid);
 
-                           local::error::reply( state, caller, common::code::xatmi::service_error);
+                           local::error::reply( state, caller, common::code::xatmi::service_error, common::process::Handle{ event.state.pid});
```
I think `state.remove(...)` should return what's needed to send error reply.
A vector of _removed-wrapper_. that holds `Caller` and `process::Handle` (callee). If anybody knows this, it's `state.remove( ...)`, after the call, the related state to the pid is gone.

lazan commented on middleware/service/source/manager/handle.cpp:
```diff
@@ -574,7 +586,7 @@ namespace casual
                               if( message.no_reply())
                                  return {};
-                              return { .process = message.process, .correlation = message.correlation, .gtrid = message.gtrid};
+                              return { .process = message.process, .correlation = message.correlation, .gtrid = message.gtrid, .service = message.requested};
```
use  `.service = service_id`

AlbsAgren committed:
service: simplify removal of reserved instances

service::manager::state::remove() now returns all the information
necessary to create a metric event for the call.

lazan commented on middleware/service/include/service/manager/state.h:
```diff
@@ -96,15 +96,28 @@ namespace casual
+               )
+            };
+
+            struct Reservation
```
Good name.

lazan reviewed (approved):
Super!

lazan commented:
Merged manually
